CONTRIBUTING.md

```diff
@@ -88,10 +88,38 @@ https://help.github.com/en/github/getting-started-with-github/git-and-github-lea
    Developers reviewing your pull request will be happy to help you add or run
    the relevant tests as part of the pull request review process.
 
-6. Commit and push your changes to your
+6. Write a useful and properly formatted commit message. 
+   Follow [these guidelines and template](https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines):
+
+     ```
+     Capitalized, short (50 chars or less) summary (no period at the end)
+     
+     More detailed explanatory text, if necessary.  Wrap it to about 72
+     characters or so.  In some contexts, the first line is treated as the
+     subject of an email and the rest of the text as the body.  The blank
+     line separating the summary from the body is critical (unless you omit
+     the body entirely); tools like rebase will confuse you if you run the
+     two together.
+     
+     Write your commit message in the imperative: "Fix bug" and not "Fixed bug"
+     or "Fixes bug."  This convention matches up with commit messages generated
+     by commands like git merge and git revert.
+     
+     Further paragraphs come after blank lines.
+     
+     - Bullet points are okay, too
+     
+     - Typically a hyphen or asterisk is used for the bullet, followed by a
+       single space, with blank lines in between, but conventions vary here
+     ```
```

Comment:
Isn't linking to the above guidelines enough?

Reply:
Maybe. I excluded the last point about a hanging indent - I think we don't need to require it. However, I wanted to quote the essential: the link points to a text that is *much more detailed*, offers context, explanation, points to further resources. I think providing a summary of the absolute essentials with a link pointing to more may be more useful than just a link. Otherwise, I fear we'll end up restating these essentials over and over in code reviews :-)

Reply:
I think it would be more effective to just point to the docs and add instead an example of a real commit message which includes for example a traceback and link to the issue fixed.

`In some contexts, the first line is treated as the
     subject of an email and the rest of the text as the body`

This is irrelevant for us.

> Otherwise, I fear we'll end up restating these essentials over and over in code reviews :-)

I think we mainly review PR titles, which end up in the release notes, but bad commit messages makes it difficult to retrospectively understand the reason for a change, so they are obviously also important.

Reply:
> I think it would be more effective to just point to the docs

But there are no *docs* per se - it's a common sense format recommended by the authors of the Git Book (and other sources). Which is why I think it's OK for us to have our own, tailored version.

> ... and add instead an example of a real commit message...

I think we could do that. Although crafting a commit message that is generic enough to serve as an example for diverse cases, but still specific enough to be useful might require considerable effort. Maybe I'm wrong.

> `In some contexts, the first line is treated as the subject of an email and the rest of the text as the body`
> 
> This is irrelevant for us.

Agreed! We can drop that line; we quote this as "adapted from...", so we can edit it.

> I think we mainly review PR titles, which end up in the release notes, but bad commit messages makes it difficult to retrospectively understand the reason for a change, so they are obviously also important.

100% agree with that - it's my primary motivation here.

Reply:
> > ... and add instead an example of a real commit message...
> 
> I think we could do that. Although crafting a commit message that is generic enough to serve as an example for diverse cases, but still specific enough to be useful might require considerable effort. Maybe I'm wrong.

Something like https://github.com/galaxyproject/galaxy/commit/0429c4d515536f9cca6b70b2abeb019de807c955 ? Not that I wanted one of mine to be used, but it was easier to search in stuff I did.

Reply:
Yes, I think that's a good example. Let's use it. Do you mind making the change?

Also, I still prefer we keep the essentials as opposed to just having a link. We could trim them to the bare minimum if what we have now is too long (I think we could have that + your example, but I'm OK either way). If we were to trim the text, I'd opt  for leaving the fist line ("Capitalized, short..."), a line from the next paragraph ("More detailed explanatory text, if necessary.  Wrap it to about 72 characters or so."), and I would definitely mention imperative mode. I feel those three items are the most significant points that apply to any commit, which is why I'd like to see them explicitly stated. But I'm OK if you really prefer the link + example with no further text.

Reply:
@nsoranzo would you like me to make this change? I think it would be good to have this in before tomorrow (paper cuts) - might be helpful to new contributors.

Reply:
Sorry, this fell through the cracks! I'll try to do that now.

Reply:
Great! Thank you - and no worries!